### ISSUE TITLE
[C#] Latest nightly C# code gen produces awful looking code

### ISSUE THREAD

**Issue by stankovski** [Azure/autorest#1380]
From the inception, the main differentiator for AutoRest (as opposed to other code generators) has been that it produced _beautiful_ code that always compiled and looked like it was hand written. Unfortunately the latest changes in the C# code generator go completely against this paradigm. 

``` cs
// Construct URL
var _baseUrl = this.BaseUri.AbsoluteUri;
var _url = new System.Uri(new System.Uri(_baseUrl + (_baseUrl.EndsWith("/") ? "" : "/")), "status").ToString();
// Create HTTP transport objects
System.Net.Http.HttpRequestMessage _httpRequest = new System.Net.Http.HttpRequestMessage();
System.Net.Http.HttpResponseMessage _httpResponse = null;
_httpRequest.Method = new System.Net.Http.HttpMethod("GET");
_httpRequest.RequestUri = new System.Uri(_url); 
```

Please bring back "usings" and find a better way to avoid type conflicts.

**Comment by brodyberg** [comment]
Can we add this goal to the documentation for the project?

**Comment by fearthecowboy** [comment]
This change was necessary to avoid some name collisions that were affecting some users.

It does have the side effect of making code that's a bit more verbose than I'd like, but changes coming in the internals of AutoRest will make it easier to insert a step where we can use Roslyn in a post-generation step to clean up some things like this, without introducing collisions.

In this *particular* case, we should be using `var` anyway. Explicit type definitions are so 2001 :D

**Comment by stankovski** [comment]
I understand the trade-off here, but this change has a negative affect on 100% of users (who generate in C#) only to fix the rare issue affecting 1-2% of scenarios. I hope the Roslyn change is coming soon.

**Comment by tbombach** [comment]
I do think it's pretty important for us to generate compilable code as much as possible. It's not entirely fair to require our users to avoid C# built-in types (they may not be C# developers, they just want a C# SDK) and especially unfair to require them to avoid our client runtime types for their model names.

Using simplified names whenever possible in the code is pretty important too so that we generate code that looks handwritten. An option we have is to use `[ModelsNamespace].[ModelName]` every time we reference a model type (though there are still possible collisions with `System.Linq` and `Newtonsoft.Json`). This would mostly prevent collisions, but I think eventually we would want to simplify those as well.

I think doing a Roslyn post-generation step is about 2 days of work (plus whatever work to integrate with Garrett's upcoming changes, though they should help make it clearer where this post-gen step fits in). I created a quick console prototype just now that loads all the generated files in a directory, adds references to the client runtimes, then uses Roslyn to simplify names. If there's a conflict, it won't simplify it. Hopefully this proposed solution would address everyone's concerns (and solve the collision issue in a more robust way). I've attached the two files to show how minimal this code would be: 
[Program.txt](https://github.com/Azure/autorest/files/446139/Program.txt)
[SimplifyNamesAnnotationRewriter.txt](https://github.com/Azure/autorest/files/446140/SimplifyNamesAnnotationRewriter.txt)
[CodeSimplifier.zip](https://github.com/Azure/autorest/files/446145/CodeSimplifier.zip).


Once Garrett's changes make it to master, I can work on integrating this code. The only other step would be adding back the `usings` directives to our templates/view models so that Roslyn knows what can be simplified. Another consideration is that it would involve shipping the C# client runtime DLLs with AutoRest, since we need those to know about those types if we want to simplify `Microsoft.Rest.*` qualified names.

Adding the explicit, fully qualified names is a necessary change for this post-gen step. This lets us disambiguate collisions if they occur. I [added a PR](https://github.com/Azure/autorest/pull/1386) for a spec that has collisions with a generated model type. That way we can make sure any future changes take this scenario into account.

**Comment by markcowl** [comment]
This is making things less readable in reviews, and will make it slightly more difficult for contributors.  If we're going to continue to be an open source library, it's important that we have readable code and that we don't introduce new StyleCop warnings without a good reason.

**Comment by fearthecowboy** [comment]
Agreed. We'll prioritize the fix.

**Comment by olydis** [comment]
fixed as of #1617
